### PR TITLE
feat: proposal for simple graphing module

### PR DESCRIPTION
**Pull request by JorgeLGonzalez**

### CODE REVIEW

File: src/graphing.module.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import {
+  SpelunkedEdge,
+  SpelunkedNode,
+  SpelunkedTree,
+} from './spelunker.interface';
+
+export class GraphingModule {
+  static graph(tree: SpelunkedTree[]): SpelunkedNode {
```
Comment (JorgeLGonzalez) [inline src/graphing.module.ts:8]: The key point is to generate a more graph-like data structure like the proposed SpelunkedNode, which is agnostic of any specific graphing renderer, but makes it much easier to traverse the graph for whatever needs you might have.

File: src/graphing.module.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }),
+      new Map<string, SpelunkedNode>(),
+    );
+
+    nodeMap.forEach((node) => this.findDependencies(node, nodeMap));
+
+    return this.findRoot(nodeMap);
+  }
+
+  static getEdges(root: SpelunkedNode): SpelunkedEdge[] {
```
Comment (JorgeLGonzalez) [inline src/graphing.module.ts:24]: This is also agnostic to any graphing lib and just returns another generic graph-like data structure the represents an edge between 2 nodes.

File: test/large-app/app.e2e-spec.ts [hunk 3]
```diff
@@ -37,3 +41,13 @@ SpelunkerSuite('Should allow the SpelunkerModule to debug', async () => {
   const output = await SpelunkerModule.debug(AppModule);
   equal(output, debugOutput);
 });
+
+SpelunkerSuite('Should allow the SpelunkerModule to graph', ({ app }) => {
+  const tree = SpelunkerModule.explore(app);
+  const root = SpelunkerModule.graph(tree);
+  const edges = SpelunkerModule.findGraphEdges(root);
+  equal(
+    edges.map((e) => `${e.from.module.name}-->${e.to.module.name}`),
```
Comment (JorgeLGonzalez) [inline test/large-app/app.e2e-spec.ts:50]: fwiw: this is the core of the mermaid syntax, but thats not really the point of findGraphEdges. The SpelunkedNode and SpelunkedEdge data structures can be easily projected to whatever syntax you want.

**Comment by jmcdo29** [comment]
Overall, code looks clean and easy to follow. Finding a better way to stop circular deps would be good, but I think what's in place would be fine as a first pass.

If you would, could you add a section to the README and add a changeset for this as a `minor` change? Just run `pnpm exec changeset` and follow the wizard.

**Comment by JorgeLGonzalez** [comment]
Ok, added to README and the changeset. Haven't used changeset before, so good chance it's not what you'd want.  I'll take a look at the circular ref issue and submit a separate PR when I get a chance.

**Comment by JorgeLGonzalez** [comment]
Actually, I added the fix for the circular dependencies.

**Review by jmcdo29 — APPROVED** [review]
All looks good to me! Thanks for adding the info in the readme. Changeset looks good. I'll get this merged and published

**Comment by jmcdo29** [comment]
Looks like the `Array.at()` method is a node 16 thing? I can update the CI to not use 12 or 14 for the tests. This isn't in the main code anyways, so I don't see the failure as a huge issue right now.

**Pull request merged**